### PR TITLE
Fix `hostdevice_vector::subspan`

### PR DESCRIPTION
**Pull request by ttnghia**
This fixes a bug in `hostdevice_vector::subspan` that incorrectly constructs a `hostdevice_span` with the input parameters.

In particular, instead of passing the `count` parameter to `hostdevice_span` constructor, it (incorrectly) subtracts `offset` from `count`.

### CODE REVIEW

**Review by davidwendt — APPROVED** [review]

File: cpp/src/io/utilities/hostdevice_vector.hpp [hunk 1]
```diff
@@ -164,13 +164,15 @@ class hostdevice_vector {
...
    *
-   * @return A typed hostdevice_span of the hostdevice_vector's data.
+   * @param offset The offset of the first element in the subspan
+   * @param count The number of elements in the subspan
+   * @return A typed hostdevice_span of the hostdevice_vector's data
    */
   [[nodiscard]] hostdevice_span<T> subspan(size_t offset, size_t count)
   {
     CUDF_EXPECTS(count >= offset, "End index cannot be smaller than the starting index.");
     CUDF_EXPECTS(count <= d_data.size(), "Slice range out of bounds.");
```
Comment (PointKernel) [inline cpp/src/io/utilities/hostdevice_vector.hpp:174]: The naming of `count` is a bit misleading. According to the above code, it seems `offset` indicates the starting index and `count` refers to the end index. So `count - offset` should be correct. @ttnghia does your change break any existing unit tests?

Comment (ttnghia) [reply cpp/src/io/utilities/hostdevice_vector.hpp:174]: No there were not any relevant unit tests for that function. I just added tests for it.

Oh these checks are also invalid. I'm going to update them.

**Review by PointKernel — APPROVED** [review]

**Review by bdice — APPROVED** [review]
Looks like a proper fix. Part of me wishes `subspan` was a free function rather than a method, so that `hostdevice_vector` would be closer in API to `std::vector`. If it were a method, I think it could be written for any container with `hostdevice` semantics (`hostdevice_vector` or `hostdevice_span`).

edit: I see there's a method `std::span<T>::subspan`, and this is based on that but applied to `hostdevice_vector`. This is taking the wrong lesson about container design from the STL, in my view, where a method could have been avoided in favor of a free function accepting any span-like object and returning a subspan that is a non-owning view subspan.

**Pull request merged**